File: src/missing.ts
import * as fs from "fs";
import { REPO } from ".";
import { Manifest, ManifestOverride } from "./manifest";
import { WikiGameCache } from "./wiki";

export function saveMissingGames(cache: WikiGameCache, manifest: Manifest, override: ManifestOverride): void {
    fs.writeFileSync(
        `${REPO}/data/missing.md`,
        Object.entries(cache)
            .sort((x, y) => x[0].localeCompare(y[0]))
            .filter(([k, _]) => (manifest[k]?.files ?? []).length === 0 && (manifest[k]?.registry ?? []).length === 0)
            .filter(([k, _]) => override[k]?.omit !== true)
            .map(([k, v]) => `* [${k}](https://www.pcgamingwiki.com/wiki/?curid=${v.pageId})`)
            .join("\n") + "\n",
    );
}

File: src/missing.rs
use itertools::Itertools;

use crate::{
    manifest::{Manifest, ManifestOverride},
    wiki::WikiCache,
    REPO,
};

pub fn save_missing_games(wiki_cache: &WikiCache, manifest: &Manifest, overrides: &ManifestOverride) {
    let lines: Vec<String> = wiki_cache
        .0
        .iter()
        .sorted_by(|(k1, _), (k2, _)| k1.to_lowercase().cmp(&k2.to_lowercase()))
        .filter(|(k, _)| {
            manifest
                .0
                .get(*k)
                .map(|x| x.files.is_empty() && x.registry.is_empty())
                .unwrap_or(true)
        })
        .filter(|(k, _)| overrides.0.get(*k).map(|x| !x.omit).unwrap_or(true))
        .map(|(k, v)| format!("* [{}](https://www.pcgamingwiki.com/wiki/?curid={})", k, v.page_id))
        .collect();

    _ = std::fs::write(
        format!("{}/data/missing.md", REPO),
        if lines.is_empty() {
            "N/A".to_string()
        } else {
            lines.join("\n") + "\n"
        },
    );
}
